network/native_endpointclient_linux.go

```diff
@@ -0,0 +1,438 @@
+	virtualGwIP, virtualGwNet, _ := net.ParseCIDR(virtualGwIPString)
+	routeInfo := RouteInfo{
+		Dst:   *virtualGwNet,
+		Scope: netlink.RT_SCOPE_LINK,
+	}
+	// Difference between interface name in addRoutes and DevName: in RouteInfo?
+	if err := addRoutes(client.netlink, client.netioshim, linkToName, []RouteInfo{routeInfo}); err != nil {
+		return err
+	}
+
+	// Add default route (ip route add default via 169.254.1.1 dev eth0)
+	_, defaultIPNet, _ := net.ParseCIDR(defaultGwCidr)
+	dstIP := net.IPNet{IP: net.ParseIP(defaultGw), Mask: defaultIPNet.Mask}
+	routeInfo = RouteInfo{
+		Dst: dstIP,
+		Gw:  virtualGwIP,
+	}
+
+	if err := addRoutes(client.netlink, client.netioshim, linkToName, []RouteInfo{routeInfo}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Helper that creates arp entry for the current NS which maps the virtual
+// gateway (169.254.1.1) to destMac on a particular interfaceName
+// Example: (169.254.1.1) at 12:34:56:78:9a:bc [ether] PERM on <interfaceName>
+func (client *NativeEndpointClient) AddDefaultArp(interfaceName string, destMac string) error {
+	_, virtualGwNet, _ := net.ParseCIDR(virtualGwIPString)
+	log.Printf("[net] Adding static arp for IP address %v and MAC %v in namespace",
+		virtualGwNet.String(), destMac)
+	hardwareAddr, err := net.ParseMAC(destMac)
+	if err != nil {
+		return err
+	}
+	if err := client.netlink.AddOrRemoveStaticArp(netlink.ADD,
+		interfaceName,
+		virtualGwNet.IP,
+		hardwareAddr,
+		false); err != nil {
+		return fmt.Errorf("adding arp entry failed: %w", err)
+	}
+	return nil
+}
+func (client *NativeEndpointClient) DeleteEndpoints(ep *endpoint) error {
+	return ExecuteInNS(client.vnetNSName, ep, client.DeleteEndpointsImpl)
+}
+func (client *NativeEndpointClient) DeleteEndpointsImpl(ep *endpoint) error {
+	routeInfoList := client.GetVnetRoutes(ep.IPAddresses)
+	if err := deleteRoutes(client.netlink, client.netioshim, client.vnetVethName, routeInfoList); err != nil {
+		return newErrorNativeEndpointClient("Failed to remove routes", err.Error())
+	}
+
+	routes, err := vishnetlink.RouteList(nil, vishnetlink.FAMILY_V4)
+	if err != nil {
+		return newErrorNativeEndpointClient("Failed to get route list", err.Error())
+	}
+	log.Printf("[native] There are %d routes remaining: %v", len(routes), routes)
+	if len(routes) <= numDefaultRoutes {
+		// Deletes default arp, default routes, ethX; there are two default routes
+		// so when we have <= numDefaultRoutes routes left, no containers use this namespace
+		log.Printf("[native] Deleting namespace %s as no containers occupy it", client.vnetNSName)
+		delErr := client.netnsClient.DeleteNamed(client.vnetNSName)
+		if delErr != nil {
+			return newErrorNativeEndpointClient("Failed to delete namespace", delErr.Error())
+		}
+	}
+	return nil
+}
+
+// Helper function that allows executing a function with one parameter in a VM namespace
+// Does not work for process namespaces
+func ExecuteInNS[T any](nsName string, param *T, f func(param *T) error) error {
```

Comment:
generics are cool, but I don't think they're necessary here. This function should accept a parameter-less function which is closed over the inputs that it needs.
For example, you could close any arbitrary context and execute it in the namespace like:
```go
func ExecuteInNS(ns string, fn func() error) error {
	//<...enter ns...>
	return fn()
}

// close the input func over arg "param"
param := "some input to my anonymous func"
ExecuteInNS("abcnamespace", func() error {
	fmt.Printf("closed over param %s", param)
	//<...do things...>
	return nil
})
```